src/hotspot/cpu/s390/stubGenerator_s390.cpp

```diff
@@ -2857,6 +2858,48 @@ class StubGenerator: public StubCodeGenerator {
     return start;
   }
 
+  address generate_nmethod_entry_barrier() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "nmethod_entry_barrier");
+
+    address start = __ pc();
+
+    // Save caller's sp & return_pc
+    __ push_frame(frame::z_abi_16_size);
+    __ z_stmg(Z_R14, Z_R15, _z_abi16(callers_sp), Z_SP);
+
+    // Prep arg for VM call
+    // Since we call address* -> int and not address -> int, we must provide an extra level of indirection.
+    // We construct a pointer to the location of R14 stored above.
+    __ z_xgr(Z_R2, Z_R2);
+    __ z_ag(Z_R2, _z_abi(return_pc), 0, Z_SP);
```

Comment:
Please use a better fitting instruction like z_la.

Reply:
The code as written here would load the contents of the stack location reserved for the return pc. Less complicated: would load the return pc. If you want to load the address of the storage location, you would use 
`z_la (Z_R2, _z_abi(return_pc), Z_R0, Z_SP);`
Load Address in it's various forms just does the address calculation, it does not access storage.

And please, use Z_R0 (not just 0) to specify an optional, unspecified register.